##  Contents
### 1.lesson Exercices
####  1.1 Import Libraries
####  1.2 Import "ords_prods_merge" dataframe
####  1.3 Create a subset
####  1.4 Define price_label function using If-Statements
####  1.5 Second Method using loc()
####  1.6 Apply the second method for "ords_prods_merge" dataframe
####  1.7 Find how busy each day is
####  1.8 Create new column "busiest day" using for-loop
### 2. Task
####  Step 2
####  Step 3
####  Step 4
####  Step 5
####  Step 7


# 1. Lesson Exercices

## 1.1 Import Libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

## 1.2 Import "ords_prods_merge" dataframe 

In [2]:
path= r'C:\Users\Bilel\Desktop\data_analyst_Carrer Foundry\Data Immersion\07-12-24 Instacart Basket Analysis'
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

## 1.3 Create a subset

In [3]:
#Create a subset for the first one million rows from "df_ords_prods_merge"
df=df_ords_prods_merge [:1000000]
df.shape

(1000000, 14)

## 1.4 Define price_label function using If-Statements

In [4]:
#define a function

def price_label (row):
    if row['prices']<= 5:
        return 'Low-range product'
    elif (row['prices']>5) and (row['prices']<= 15):
         return 'Mid-range product'
    elif row ['prices']>15:
        return 'High-range product'
    else: return 'Not enough data'
        


In [5]:
# Apply the function
df['price_range']= df.apply(price_label, axis=1)

C:\Users\Bilel\AppData\Local\Temp\ipykernel_16236\4155366810.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range']= df.apply(price_label, axis=1)


In [6]:
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product     673183
Low-range product     314392
High-range product     12425
Name: count, dtype: int64

In [7]:
df['prices'].max()

99999.0

In [8]:
df['prices'].min()

1.0

## 1.5 Second Method using loc()

In [9]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


C:\Users\Bilel\AppData\Local\Temp\ipykernel_16236\2857404532.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [10]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [11]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [12]:
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     673183
Low-range product     314392
High-range product     12425
Name: count, dtype: int64

## 1.6 Apply the second method for "ords_prods_merge" dataframe

In [13]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] ="High-range product"

In [14]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [15]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] ="Low-range product"

In [16]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

## 1.7 Find how busy each day is 

In [17]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

## 1.8 Create new column "busiest day" using for-loop

In [18]:
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [19]:
df_ords_prods_merge['busiest_day']= result

In [20]:
df_ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 2. TasK

## Step 2 : Create a new column that lables the 2 "Busiest days" and 2 "Slowest days"

In [21]:
# Find busiest and slowest days
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)


orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [22]:
# Days 0 and 1 (Saturday and Sunday) are the busiest days , Days 3 and 4 (Thursday and Wednesday) are solwest days

In [23]:
result_1 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
    if value == 0 or value == 1:
        result_1.append("Busiest day")
    elif value == 3 or value == 4:
        result_1.append("Slowest day")
    else:
        result_1.append("Regular day")


In [24]:
result_1

['Regular day',
 'Regular day',
 'Regular day',
 'Regular day',
 'Regular day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Regular day',
 'Regular day',
 'Regular day',
 'Regular day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Slowest day',
 'Regular day',
 'Regular day',
 'Regular day',
 'Regula

In [25]:
df_ords_prods_merge['Busiest_days']= result_1


## Step 3 : Check values of new "Busiest days" column for accuracy

In [26]:
df_ords_prods_merge['Busiest_days'].value_counts(dropna=False)

Busiest_days
Regular day    12916111
Busiest day    11864412
Slowest day     7624336
Name: count, dtype: int64

In [27]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc     busiest_day Busiest_days  
0              7     9.0  Mid-range product  Regularly busy  Regular day  
1             16    12.5  Mid-range product  Regularly busy  Regular day  
2             19     4.4  Low-range product  Regularly busy  Regular day  
3             19     4.7  Low-range product  Regularly busy  Regular day  
4             17     1.0  Low-range product  Regularly busy  Regular day

## <span style="color:green"> observation 
The total value counts for the df_ords_prods_merge dataframe is equal to the total sum of all groupings . In addition, the listed sums for each label match the sum totals for the qualifying days of the week (e.g. Total "Busiest day" = 11864412, which matches with th total of "orders_day_of_week" 0 & 1 (6204182 + 5660230)).


## Step 4 : Create new column "busiest_period_of_day"

In [28]:
# Check value counts in "order_hour_of_day" column.
df_ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

## <span style="color:green"> observation 
The value counts listed above will be split into equal thirds for the following labels: "Most orders" = 10, 11, 14, 15, 13, 12, 16, and 9; "Fewest orders" = 23, 6, 0, 1, 5, 2, 4, and 3; "Average orders" = all remaining values in order_hour_of_day.

In [29]:
# Create for-loop if statement labeling periods of time as “Most orders,” “Average orders,” and “Fewest orders.”

result_2 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_2.append("Most orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_2.append("Fewest orders")
  else:
    result_2.append("Average orders")
    

In [30]:
result_2

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Avera

In [31]:
# Create new column "busiest_period_of_day" in ords_prods_merge.
df_ords_prods_merge['busiest_period_of_day'] = result_2

## Step 5 :  Print frequency of "busiest_period_of_day" column

In [32]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

## Step 7 :  Export dataframe as a pickle file to “Prepared Data” folder

In [33]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_derived.pkl'))